demos/common/utilities/jws_using_mbedtls.c

```diff
@@ -0,0 +1,593 @@
+                                             pucE, ulELength );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "mbedtls_rsa_import_raw res: %i\n", lMbedTLSResult ) );
+    }
+
+    lMbedTLSResult = mbedtls_rsa_complete( &ctx );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "mbedtls_rsa_complete res: %i\n", lMbedTLSResult ) );
+    }
+
+    lMbedTLSResult = mbedtls_rsa_check_pubkey( &ctx );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "mbedtls_rsa_check_pubkey res: %i\n", lMbedTLSResult ) );
+    }
+
+    LogInfo( ( "--- Decrypting ---\n" ) );
+
+    /* RSA */
+    lMbedTLSResult = mbedtls_rsa_pkcs1_decrypt( &ctx, NULL, NULL, MBEDTLS_RSA_PUBLIC, &ulDecryptedLength, pucSignature, pucBuffer, jwsRSA3072_SIZE );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "mbedtls_rsa_pkcs1_decrypt res: %i\n", lMbedTLSResult ) );
+    }
+
+    LogInfo( ( "---- Calculating SHA256 over input ----\n" ) );
+    xResult = prvJWS_SHA256Calculate( pucInput, ulInputLength,
+                                      pucShaBuffer );
+
+    LogInfo( ( "--- Checking if SHA256 of header+payload matches decrypted SHA256 ---\n" ) );
+
+    /* TODO: remove this once we have a valid PKCS7 parser. */
+    int doTheyMatch = memcmp( pucBuffer + jwsPKCS7_PAYLOAD_OFFSET, pucShaBuffer, jwsSHA256_SIZE );
+
+    if( doTheyMatch == 0 )
+    {
+        LogInfo( ( "SHA of JWK matches\n" ) );
+        xResult = 0;
+    }
+    else
+    {
+        LogError( ( "SHA of JWK does NOT match\n" ) );
+        xResult = 1;
+    }
+
+    return xResult;
+}
+
+
+uint32_t JWS_ManifestVerify( const char * pucManifest,
+                             uint32_t ulManifestLength,
+                             char * pucJWS,
+                             uint32_t ulJWSLength )
+{
+    uint32_t ulVerificationResult;
+
+    int lMbedResult;
+    unsigned char * pucHeader;
+    unsigned char * pucPayload;
+    unsigned char * pucSignature;
+    uint32_t ulHeaderLength;
+    uint32_t ulPayloadLength;
+    uint32_t ulSignatureLength;
+    AzureIoTJSONReader_t xJSONReader;
+
+    LogInfo( ( "---------------------Begin Signature Validation --------------------\n\n" ) );
+
+    /*------------------- Parse and Decode the Manifest Sig ------------------------*/
+
+    AzureIoTResult_t xResult = prvSplitJWS( ( unsigned char * ) pucJWS, ulJWSLength,
+                                            &pucHeader, &ulHeaderLength,
+                                            &pucPayload, &ulPayloadLength,
+                                            &pucSignature, &ulSignatureLength );
+    prvSwapToUrlEncodingChars( pucSignature, ulSignatureLength );
+
+    LogInfo( ( "---JWS Base64 Decode Header---\n" ) );
+    int32_t outJWSDecodedHeaderLength;
+    lMbedResult = mbedtls_base64_decode( ( unsigned char * ) ucBase64DecodedHeader, sizeof( ucBase64DecodedHeader ), ( size_t * ) &outJWSDecodedHeaderLength, ( const unsigned char * ) pucHeader, ulHeaderLength );
```

Comment:
ugh. after decoding ucBase64DecodedHeader is nothing special, just a buffer, while is actually encoded. I would swap the tag around as in:

pucHeader -> pucBase64EncodedHeader
ucBase64DecodedHeader ->ucHeader